pkg/scheduler/internal/queue/scheduling_queue.go

```diff
@@ -888,20 +892,34 @@ func (u *UnschedulablePods) addOrUpdate(pInfo *framework.QueuedPodInfo) {
 	u.podInfoMap[podID] = pInfo
 }
 
-// Delete deletes a pod from the unschedulable podInfoMap.
-func (u *UnschedulablePods) delete(pInfo *framework.QueuedPodInfo) {
+// delete deletes a podInfo from the unschedulable podInfoMap.
+// The `gated` parameter is used to figure out which metric should be decreased.
+func (u *UnschedulablePods) delete(pInfo *framework.QueuedPodInfo, gated bool) {
 	podID := u.keyFunc(pInfo.Pod)
 	if _, exists := u.podInfoMap[podID]; exists {
-		if pInfo.Gated && u.gatedRecorder != nil {
+		if gated && u.gatedRecorder != nil {
 			u.gatedRecorder.Dec()
-		} else if !pInfo.Gated && u.unschedulableRecorder != nil {
+		} else if !gated && u.unschedulableRecorder != nil {
+			u.unschedulableRecorder.Dec()
+		}
+	}
+	delete(u.podInfoMap, podID)
+}
+
+// deletePod deletes a pod from the unschedulable podInfoMap.
+func (u *UnschedulablePods) deletePod(pod *v1.Pod) {
```

Comment:
why not just use this function everywhere? why do we need the other one?

Reply:
Because addToActiveQ() modifies the pInfo.gated in place, we need to pass the `gated` value in some cases like:

```go
	gated := pInfo.Gated
	if added, _ := p.addToActiveQ(pInfo); !added {
		return false
	}
	p.unschedulablePods.delete(pInfo, gated)
```

Reply:
Tracking the metric on each addition/deletion is concerning because it is easy to get wrong; ideally we should update the count (like ```metric.Set(len(podInfoMap))```); however, we don't have a separate map for the gated pods, so we are stuck with this approach for now I guess :(

But I still prefer to not have two functions, I made a suggestion how we can potentially do that.

Reply:
> however, we don't have a separate map for the gated pods, so we are stuck with this approach for now I guess :(

indeed, an ideal way is just like you said, count the internal queue size and use it as the source of truth, call `metric.Set()` by the end of each operation to sync the metrics.

This will involve some non-trivial changes, but worth a follow-up.